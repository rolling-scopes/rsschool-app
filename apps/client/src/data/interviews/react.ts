import { InterviewTemplate } from './types';

export const reactTemplate: InterviewTemplate = {
  name: 'React interview ',
  examplesUrl: 'https://github.com/rolling-scopes-school/tasks/blob/master/react/questions.md',
  categories: [
    {
      id: 1000,
      name: 'JSX',
      questions: [
        { id: 1001, name: 'What is JSX?' },
        { id: 1002, name: 'Is it possible to use React without JSX?' },
      ],
    },
    {
      id: 1010,
      name: 'React DOM',
      questions: [
        { id: 1011, name: 'What is the virtual DOM? How does react use the virtual DOM to render the UI?' },
        { id: 1012, name: 'Is the virtual DOM the same as the shadow DOM?' },
        { id: 1013, name: 'What is the difference between the virtual DOM and the real DOM?' },
      ],
    },
    {
      id: 1020,
      name: 'Render',
      questions: [
        { id: 1021, name: 'When is a component rendered?' },
        { id: 1022, name: 'How not to render on props change?' },
        { id: 1023, name: 'Is it OK to use arrow functions in render methods?' },
      ],
    },
    {
      id: 1030,
      name: 'Interaction between components',
      questions: [
        { id: 1031, name: 'How do you pass a value from parent to child?' },
        { id: 1032, name: 'How do you pass a value from child to parent?' },
        { id: 1033, name: 'What is prop drilling?' },
        { id: 1034, name: 'Can a child component modify its own props?' },
        { id: 1035, name: 'How do you pass a value from sibling to sibling?' },
      ],
    },
    {
      id: 1040,
      name: 'Lifecycle and State',
      questions: [
        { id: 1041, name: 'What is the difference between props and state?' },
        { id: 1042, name: 'How does state in a class component differ from state in a functional component?' },
        { id: 1043, name: 'What is the component lifecycle?' },
        { id: 1044, name: 'How do you update lifecycle in function components?' },
      ],
    },
    {
      id: 1050,
      name: 'Ref',
      questions: [
        { id: 1051, name: 'What is the difference between refs and state variables?' },
        { id: 1052, name: 'When is the best time to use refs?' },
        { id: 1053, name: 'What is the proper way to update a ref in a function component?' },
      ],
    },
    {
      id: 1060,
      name: 'Context',
      questions: [
        { id: 1061, name: 'What is the difference between the context API and prop drilling?' },
        { id: 1062, name: "When shouldn't you use the context API?" },
      ],
    },
    {
      id: 1070,
      name: 'Redux',
      questions: [
        { id: 1071, name: 'Enumerate base principles' },
        { id: 1072, name: 'What is the typical flow of data in a React + Redux app?' },
        { id: 1073, name: 'Benefits of Redux?' },
      ],
    },
    {
      id: 1080,
      name: 'Other',
      questions: [
        { id: 1081, name: 'Is it a good idea to use Math.random for keys?' },
        { id: 1082, name: 'What are the limitations of React?' },
        { id: 1083, name: 'What is a higher order component?' },
        { id: 1084, name: 'What are uncontrolled and controlled components?' },
        { id: 1085, name: 'React optimizations' },
      ],
    },
    {
      id: 1100,
      name: 'Coding task',
      questions: [{ id: 1101, name: 'Small react app: form, button, results list' }],
    },
  ],
};
