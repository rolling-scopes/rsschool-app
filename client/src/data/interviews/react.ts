import { InterviewTemplate } from './types';

export const reactTemplate: InterviewTemplate = {
  name: 'React interview',
  examplesUrl: 'https://github.com/rolling-scopes-school/tasks/blob/master/react/questions.md',
  categories: [
    {
      id: 1000,
      name: 'JSX',
      questions: [
        { id: 1001, name: 'What is JSX?' },
        { id: 1002, name: 'Is it possible to use React without JSX?' },
      ],
    },
    {
      id: 1010,
      name: 'Styled-components',
      questions: [{ id: 1011, name: 'Using styled-components' }],
    },
    {
      id: 1020,
      name: 'Components',
      questions: [
        { id: 1021, name: 'Component Definition (Class-based, Functional)' },
        { id: 1022, name: 'Components Composition' },
      ],
    },
    {
      id: 1030,
      name: 'React DOM',
      questions: [
        { id: 1031, name: 'What is the virtual DOM? How does react use the virtual DOM to render the UI?' },
        { id: 1032, name: 'Is the virtual DOM the same as the shadow DOM?' },
        { id: 1033, name: 'What is the difference between the virtual DOM and the real DOM?' },
      ],
    },
    {
      id: 1040,
      name: 'Render',
      questions: [
        { id: 1041, name: 'When is a component rendered?' },
        { id: 1042, name: 'How not to render on props change?' },
        { id: 1043, name: 'Is it OK to use arrow functions in render methods?' },
        { id: 1044, name: 'Components rendering system' },
      ],
    },
    {
      id: 1050,
      name: 'Interaction between components',
      questions: [
        { id: 1051, name: 'How do you pass a value from parent to child?' },
        { id: 1052, name: 'How do you pass a value from child to parent?' },
        { id: 1053, name: 'What is prop drilling?' },
        { id: 1054, name: 'Can a child component modify its own props?' },
        { id: 1055, name: 'How do you pass a value from sibling to sibling?' },
      ],
    },
    {
      id: 1060,
      name: 'Lifecycle and State',
      questions: [
        { id: 1061, name: 'What is the difference between props and state?' },
        { id: 1062, name: 'How does state in a class component differ from state in a functional component?' },
        { id: 1063, name: 'What is the component lifecycle?' },
        { id: 1064, name: 'How do you update lifecycle in function components?' },
        { id: 1065, name: 'Controlled/uncontrolled components' },
        { id: 1066, name: 'Stateful vs stateless components' },
      ],
    },
    {
      id: 1070,
      name: 'Ref',
      questions: [
        { id: 1071, name: 'What is the difference between refs and state variables?' },
        { id: 1072, name: 'When is the best time to use refs?' },
        { id: 1073, name: 'What is the proper way to update a ref in a function component?' },
      ],
    },
    {
      id: 1080,
      name: 'Context',
      questions: [
        { id: 1081, name: 'What is the difference between the context API and prop drilling?' },
        { id: 1082, name: "When shouldn't you use the context API?" },
      ],
    },
    {
      id: 1090,
      name: 'Redux',
      questions: [
        { id: 1091, name: 'Enumerate base principles' },
        { id: 1092, name: 'What is the typical flow of data in a React + Redux app?' },
        { id: 1093, name: 'Benefits of Redux?' },
        { id: 1094, name: 'Async Redux flow' },
      ],
    },
    {
      id: 1100,
      name: 'State Management by managers',
      questions: [
        { id: 1101, name: 'Mobx' },
        { id: 1102, name: 'Reflux' },
      ],
    },
    {
      id: 1110,
      name: 'Routing',
      questions: [
        { id: 1111, name: 'React Router' },
        { id: 1112, name: 'History' },
      ],
    },
    {
      id: 1120,
      name: 'Other',
      questions: [
        { id: 1121, name: 'Is it a good idea to use Math.random for keys?' },
        { id: 1122, name: 'What are the limitations of React?' },
        { id: 1123, name: 'What is a higher order component?' },
        { id: 1124, name: 'What are uncontrolled and controlled components?' },
        { id: 1125, name: 'React optimizations' },
      ],
    },
    {
      id: 1130,
      name: 'Coding task',
      questions: [{ id: 1131, name: 'Small react app: form, button, results list' }],
    },
    {
      id: 1140,
      name: 'Next.js',
      questions: [
        { id: 1141, name: 'What is Next.js, and how does it differ from traditional React applications?' },
        { id: 1142, name: 'Explain the concept of server-side rendering (SSR) in Next.js' },
        { id: 1143, name: 'How does Next.js handle client-side routing?' },
        { id: 1144, name: 'What are the benefits of using Next.js for building React applications?' },
        {
          id: 1145,
          name: 'What is the purpose of the getInitialProps function in Next.js? How is it different from using getStaticProps or getServerSideProps?',
        },
        { id: 1146, name: 'How does Next.js handle automatic code splitting, and why is it important?' },
        {
          id: 1147,
          name: 'What is the purpose of the Link component in Next.js, and how does it differ from traditional anchor (<a>) tags?',
        },
        { id: 1148, name: 'How can you configure routing in a Next.js application?' },
      ],
    },
    {
      id: 1150,
      name: 'Data Fetching',
      questions: [
        {
          id: 1151,
          name: 'Compare and contrast getStaticProps and getServerSideProps in terms of use cases and performance considerations',
        },
        {
          id: 1152,
          name: 'When would you choose to use getStaticPaths in Next.js, and how does it relate to dynamic routes?',
        },
      ],
    },
    {
      id: 1160,
      name: 'Advanced Topics',
      questions: [
        { id: 1161, name: 'Describe the purpose and use cases of API routes in Next.js' },
        { id: 1162, name: 'Explain the concept of middleware in Next.js and its role in the application lifecycle' },
        { id: 1163, name: 'How does Next.js handle authentication in applications?' },
      ],
    },
  ],
};
