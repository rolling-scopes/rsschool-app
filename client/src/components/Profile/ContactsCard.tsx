import * as React from 'react';
import isEqual from 'lodash/isEqual';
import { List, Typography, Input, Form } from 'antd';
import CommonCard from './CommonCard';
import { Contacts } from 'common/models/profile';
import { ConfigurableProfilePermissions } from 'common/models/profile';
import { ChangedPermissionsSettings } from 'pages/profile';
import { CheckboxChangeEvent } from 'antd/lib/checkbox';

const { Text } = Typography;

import { ContactsOutlined } from '@ant-design/icons';
import { NotificationChannel } from 'modules/Notifications/services/notifications';
import { EmailConfirmation } from './EmailConfirmation';
import { Rule } from 'antd/lib/form';

type ProfileEntryChangeCallback = (event: { target: { value: string } }, path: string) => void;

type Props = {
  data: Contacts;
  initialContacts?: Contacts;
  isEditingModeEnabled: boolean;
  permissionsSettings?: ConfigurableProfilePermissions;
  onPermissionsSettingsChange: (event: CheckboxChangeEvent, settings: ChangedPermissionsSettings) => void;
  onProfileSettingsChange: ProfileEntryChangeCallback;
  isDataPendingSave: boolean;
  connections: Partial<
    Record<
      NotificationChannel,
      | {
          value: string;
          enabled: boolean;
          lastLinkSentAt?: string;
        }
      | undefined
    >
  >;
  sendConfirmationEmail: () => void;
};

type Contact = { name: string; value: string | null; key: string; rules?: Rule[] };

class ContactsCard extends React.Component<Props> {
  state = {
    data: null,
  };

  private filterPermissions = ({
    isEmailVisible,
    isTelegramVisible,
    isPhoneVisible,
    isSkypeVisible,
    isContactsNotesVisible,
    isLinkedInVisible,
  }: Partial<ConfigurableProfilePermissions>) => ({
    isEmailVisible,
    isTelegramVisible,
    isPhoneVisible,
    isSkypeVisible,
    isContactsNotesVisible,
    isLinkedInVisible,
  });

  shouldComponentUpdate = (nextProps: Props) => {
    const {
      isEmailVisible,
      isTelegramVisible,
      isPhoneVisible,
      isSkypeVisible,
      isContactsNotesVisible,
      isLinkedInVisible,
    } = this.props.permissionsSettings!;

    return (
      !isEqual(nextProps.data, this.props.data) ||
      !isEqual(nextProps.permissionsSettings?.isEmailVisible, isEmailVisible) ||
      !isEqual(nextProps.permissionsSettings?.isTelegramVisible, isTelegramVisible) ||
      !isEqual(nextProps.permissionsSettings?.isPhoneVisible, isPhoneVisible) ||
      !isEqual(nextProps.permissionsSettings?.isSkypeVisible, isSkypeVisible) ||
      !isEqual(nextProps.permissionsSettings?.isContactsNotesVisible, isContactsNotesVisible) ||
      !isEqual(nextProps.permissionsSettings?.isLinkedInVisible, isLinkedInVisible) ||
      !isEqual(nextProps.isEditingModeEnabled, this.props.isEditingModeEnabled) ||
      !isEqual(nextProps.connections, this.props.connections) ||
      nextProps.isDataPendingSave !== this.props.isDataPendingSave
    );
  };

  render() {
    const {
      isEditingModeEnabled,
      permissionsSettings,
      onPermissionsSettingsChange,
      onProfileSettingsChange,
      connections,
      sendConfirmationEmail,
      isDataPendingSave,
      initialContacts,
    } = this.props;
    const { email, epamEmail, telegram, phone, skype, notes, linkedIn } = this.props.data;
    const contacts: Contact[] = [
      {
        name: 'EPAM E-mail',
        value: epamEmail,
        key: 'epamEmail',
        rules: [{ type: 'email', message: 'Email is not valid' }],
      },
      {
        name: 'E-mail',
        value: email,
        key: 'email',
        rules: [{ type: 'email', message: 'Email is not valid' }],
      },
      {
        name: 'Telegram',
        value: telegram,
        key: 'telegram',
      },
      {
        name: 'Phone',
        value: phone,
        key: 'phone',
      },
      {
        name: 'Skype',
        value: skype,
        key: 'skype',
      },
      {
        name: 'Notes',
        value: notes,
        key: 'notes',
      },
      {
        name: 'LinkedIn',
        value: linkedIn,
        key: 'linkedIn',
      },
    ];

    const filledContacts = contacts.filter(({ value }: Contact) => value);

    return (
      <CommonCard
        title="Contacts"
        icon={<ContactsOutlined />}
        content={
          filledContacts.length ? (
            <List
              itemLayout="horizontal"
              dataSource={filledContacts}
              renderItem={({ name, value, key }: Contact) => (
                <List.Item>
                  <Text strong>{name}:</Text>{' '}
                  {key !== 'linkedIn' ? (
                    <>
                      {value}
                      {key === 'email' && !isDataPendingSave && (!connections.email || !connections.email.enabled) ? (
                        <EmailConfirmation
                          connection={connections.email}
                          sendConfirmationEmail={sendConfirmationEmail}
                        />
                      ) : null}
                    </>
                  ) : (
                    <a target="__blank" href={value!}>
                      {value}
                    </a>
                  )}
                </List.Item>
              )}
            />
          ) : undefined
        }
        noDataDescrption="Contacts aren't filled in"
        permissionsSettings={permissionsSettings ? this.filterPermissions(permissionsSettings) : undefined}
        isEditingModeEnabled={isEditingModeEnabled}
        onPermissionsSettingsChange={onPermissionsSettingsChange}
        detachSettingsOnVisibilityChange
        profileSettingsContent={
          <EditForm
            contacts={contacts}
            onProfileSettingsChange={onProfileSettingsChange}
            initialContacts={initialContacts}
          />
        }
      />
    );
  }
}

export default ContactsCard;

function EditForm({
  contacts,
  onProfileSettingsChange,
  initialContacts,
}: {
  contacts: Contact[];
  initialContacts?: Contacts;
  onProfileSettingsChange: ProfileEntryChangeCallback;
}) {
  const [form] = Form.useForm();
  return (
    <Form
      form={form}
      onValuesChange={(changedValues: Record<string, string>) => {
        function notifyChanges(errors?: { errorFields: { errors: string[]; name: string[] }[] }) {
          Object.entries(changedValues).forEach(([key, value]: [string, string]) => {
            const path = `contacts.${key}`;
            if (!errors || errors.errorFields.every(field => !field.name.includes(key))) {
              onProfileSettingsChange({ target: { value } }, path);
            } else if (initialContacts) {
              onProfileSettingsChange(
                { target: { value: (initialContacts as unknown as Record<string, string>)[key] ?? '' } },
                path,
              );
            }
          });
        }
        form
          .validateFields()
          .then(() => notifyChanges())
          .catch(errors => notifyChanges(errors));
      }}
      initialValues={Object.fromEntries(contacts.map(c => [c.key, c.value]))}
    >
      <List<Contact>
        itemLayout="horizontal"
        dataSource={contacts}
        renderItem={({ name, key, rules }: Contact) => (
          <List.Item>
            <div style={{ width: '100%' }}>
              <p style={{ fontSize: 18, marginBottom: 5 }}>
                <Text strong>{name}:</Text>
              </p>
              <Form.Item rules={rules} name={key}>
                <Input style={{ width: '100%' }} />
              </Form.Item>
            </div>
          </List.Item>
        )}
      />
    </Form>
  );
}
