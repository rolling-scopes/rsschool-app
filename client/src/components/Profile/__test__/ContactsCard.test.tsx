import React from 'react';
import { render } from '@testing-library/react';
import ContactsCard from '../ContactsCard';

describe('ContactsCard', () => {
  describe('Should render correctly', () => {
    it('if editing mode is disabled', () => {
      const wrapper = render(
        <ContactsCard
          data={{
            epamEmail: 'vasya@epam.com',
            phone: '1232422',
            email: 'vasya@tut.by',
            skype: 'skype_vasya',
            telegram: 'televasya',
            notes: 'vasya',
            linkedIn: 'http://linkedin_test.com/vasya',
          }}
          isEditingModeEnabled={false}
          sendConfirmationEmail={jest.fn()}
          connections={{}}
          updateProfile={jest.fn()}
        />,
      );
      expect(wrapper.container).toMatchSnapshot();
    });
    it('if editing mode is enabled', () => {
      const wrapper = render(
        <ContactsCard
          data={{
            epamEmail: 'vasya@epam.com',
            phone: '1232422',
            email: 'vasya@tut.by',
            skype: 'skype_vasya',
            telegram: null,
            notes: null,
            linkedIn: null,
          }}
          isEditingModeEnabled={true}
          sendConfirmationEmail={jest.fn()}
          connections={{}}
          updateProfile={jest.fn()}
        />,
      );
      expect(wrapper.container).toMatchSnapshot();
    });
  });
});
