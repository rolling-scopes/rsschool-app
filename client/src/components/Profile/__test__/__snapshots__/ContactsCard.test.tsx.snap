// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactsCard Should render correctly if editing mode is disabled 1`] = `
<div>
  <div
    class="ant-card ant-card-bordered css-dev-only-do-not-override-nllxry"
  >
    <div
      class="ant-card-head"
    >
      <div
        class="ant-card-head-wrapper"
      >
        <div
          class="ant-card-head-title"
        >
          <h2
            class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line css-dev-only-do-not-override-nllxry"
            style="font-size: 16px; margin-bottom: 0px; display: flex; align-items: center; justify-content: space-between;"
          >
            <span>
              <span
                aria-label="contacts"
                class="anticon anticon-contacts"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="contacts"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M594.3 601.5a111.8 111.8 0 0029.1-75.5c0-61.9-49.9-112-111.4-112s-111.4 50.1-111.4 112c0 29.1 11 55.5 29.1 75.5a158.09 158.09 0 00-74.6 126.1 8 8 0 008 8.4H407c4.2 0 7.6-3.3 7.9-7.5 3.8-50.6 46-90.5 97.2-90.5s93.4 40 97.2 90.5c.3 4.2 3.7 7.5 7.9 7.5H661a8 8 0 008-8.4c-2.8-53.3-32-99.7-74.7-126.1zM512 578c-28.5 0-51.7-23.3-51.7-52s23.2-52 51.7-52 51.7 23.3 51.7 52-23.2 52-51.7 52zm416-354H768v-56c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v56H548v-56c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v56H328v-56c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v56H96c-17.7 0-32 14.3-32 32v576c0 17.7 14.3 32 32 32h832c17.7 0 32-14.3 32-32V256c0-17.7-14.3-32-32-32zm-40 568H136V296h120v56c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-56h148v56c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-56h148v56c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-56h120v496z"
                  />
                </svg>
              </span>
               
              Contacts
            </span>
          </h2>
        </div>
      </div>
    </div>
    <div
      class="ant-card-body"
    >
      <div
        class="ant-list ant-list-split css-dev-only-do-not-override-nllxry"
      >
        <div
          class="ant-spin-nested-loading css-dev-only-do-not-override-nllxry"
        >
          <div
            class="ant-spin-container"
          >
            <ul
              class="ant-list-items"
            >
              <li
                class="ant-list-item ant-list-item-no-flex"
              >
                <span
                  class="ant-typography css-dev-only-do-not-override-nllxry"
                >
                  <strong>
                    EPAM E-mail
                    :
                  </strong>
                </span>
                 
                vasya@epam.com
              </li>
              <li
                class="ant-list-item ant-list-item-no-flex"
              >
                <span
                  class="ant-typography css-dev-only-do-not-override-nllxry"
                >
                  <strong>
                    E-mail
                    :
                  </strong>
                </span>
                 
                vasya@tut.by
              </li>
              <li
                class="ant-list-item ant-list-item-no-flex"
              >
                <span
                  class="ant-typography css-dev-only-do-not-override-nllxry"
                >
                  <strong>
                    Telegram
                    :
                  </strong>
                </span>
                 
                televasya
              </li>
              <li
                class="ant-list-item ant-list-item-no-flex"
              >
                <span
                  class="ant-typography css-dev-only-do-not-override-nllxry"
                >
                  <strong>
                    Phone
                    :
                  </strong>
                </span>
                 
                1232422
              </li>
              <li
                class="ant-list-item ant-list-item-no-flex"
              >
                <span
                  class="ant-typography css-dev-only-do-not-override-nllxry"
                >
                  <strong>
                    Skype
                    :
                  </strong>
                </span>
                 
                skype_vasya
              </li>
              <li
                class="ant-list-item ant-list-item-no-flex"
              >
                <span
                  class="ant-typography css-dev-only-do-not-override-nllxry"
                >
                  <strong>
                    Notes
                    :
                  </strong>
                </span>
                 
                vasya
              </li>
              <li
                class="ant-list-item ant-list-item-no-flex"
              >
                <span
                  class="ant-typography css-dev-only-do-not-override-nllxry"
                >
                  <strong>
                    LinkedIn
                    :
                  </strong>
                </span>
                 
                <a
                  href="http://linkedin_test.com/vasya"
                  target="__blank"
                >
                  http://linkedin_test.com/vasya
                </a>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ContactsCard Should render correctly if editing mode is enabled 1`] = `
<div>
  <div
    class="ant-card ant-card-bordered css-dev-only-do-not-override-nllxry"
  >
    <div
      class="ant-card-head"
    >
      <div
        class="ant-card-head-wrapper"
      >
        <div
          class="ant-card-head-title"
        >
          <h2
            class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line css-dev-only-do-not-override-nllxry"
            style="font-size: 16px; margin-bottom: 0px; display: flex; align-items: center; justify-content: space-between;"
          >
            <span>
              <span
                aria-label="contacts"
                class="anticon anticon-contacts"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="contacts"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M594.3 601.5a111.8 111.8 0 0029.1-75.5c0-61.9-49.9-112-111.4-112s-111.4 50.1-111.4 112c0 29.1 11 55.5 29.1 75.5a158.09 158.09 0 00-74.6 126.1 8 8 0 008 8.4H407c4.2 0 7.6-3.3 7.9-7.5 3.8-50.6 46-90.5 97.2-90.5s93.4 40 97.2 90.5c.3 4.2 3.7 7.5 7.9 7.5H661a8 8 0 008-8.4c-2.8-53.3-32-99.7-74.7-126.1zM512 578c-28.5 0-51.7-23.3-51.7-52s23.2-52 51.7-52 51.7 23.3 51.7 52-23.2 52-51.7 52zm416-354H768v-56c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v56H548v-56c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v56H328v-56c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v56H96c-17.7 0-32 14.3-32 32v576c0 17.7 14.3 32 32 32h832c17.7 0 32-14.3 32-32V256c0-17.7-14.3-32-32-32zm-40 568H136V296h120v56c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-56h148v56c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-56h148v56c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-56h120v496z"
                  />
                </svg>
              </span>
               
              Contacts
            </span>
            <span
              aria-label="edit"
              class="anticon anticon-edit"
              role="img"
              tabindex="-1"
            >
              <svg
                aria-hidden="true"
                data-icon="edit"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M257.7 752c2 0 4-.2 6-.5L431.9 722c2-.4 3.9-1.3 5.3-2.8l423.9-423.9a9.96 9.96 0 000-14.1L694.9 114.9c-1.9-1.9-4.4-2.9-7.1-2.9s-5.2 1-7.1 2.9L256.8 538.8c-1.5 1.5-2.4 3.3-2.8 5.3l-29.5 168.2a33.5 33.5 0 009.4 29.8c6.6 6.4 14.9 9.9 23.8 9.9zm67.4-174.4L687.8 215l73.3 73.3-362.7 362.6-88.9 15.7 15.6-89zM880 836H144c-17.7 0-32 14.3-32 32v36c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-36c0-17.7-14.3-32-32-32z"
                />
              </svg>
            </span>
          </h2>
        </div>
      </div>
    </div>
    <div
      class="ant-card-body"
    >
      <div
        class="ant-list ant-list-split css-dev-only-do-not-override-nllxry"
      >
        <div
          class="ant-spin-nested-loading css-dev-only-do-not-override-nllxry"
        >
          <div
            class="ant-spin-container"
          >
            <ul
              class="ant-list-items"
            >
              <li
                class="ant-list-item ant-list-item-no-flex"
              >
                <span
                  class="ant-typography css-dev-only-do-not-override-nllxry"
                >
                  <strong>
                    EPAM E-mail
                    :
                  </strong>
                </span>
                 
                vasya@epam.com
              </li>
              <li
                class="ant-list-item ant-list-item-no-flex"
              >
                <span
                  class="ant-typography css-dev-only-do-not-override-nllxry"
                >
                  <strong>
                    E-mail
                    :
                  </strong>
                </span>
                 
                vasya@tut.by
                <div
                  class="ant-alert ant-alert-error ant-alert-no-icon css-dev-only-do-not-override-nllxry"
                  data-show="true"
                  role="alert"
                >
                  <div
                    class="ant-alert-content"
                  >
                    <div
                      class="ant-alert-message"
                    >
                      <div>
                        Email is not verified.
                         
                        <span
                          style="text-decoration: underline; cursor: pointer;"
                        >
                          Send confirmation email?
                        </span>
                      </div>
                    </div>
                  </div>
                </div>
              </li>
              <li
                class="ant-list-item ant-list-item-no-flex"
              >
                <span
                  class="ant-typography css-dev-only-do-not-override-nllxry"
                >
                  <strong>
                    Phone
                    :
                  </strong>
                </span>
                 
                1232422
              </li>
              <li
                class="ant-list-item ant-list-item-no-flex"
              >
                <span
                  class="ant-typography css-dev-only-do-not-override-nllxry"
                >
                  <strong>
                    Skype
                    :
                  </strong>
                </span>
                 
                skype_vasya
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
